data_structures/binary_search_tree.cpp

```diff
@@ -1,172 +1,296 @@
 /**
  * \file
  * \brief A simple tree implementation using structured nodes
- *
- * \todo update code to use C++ STL library features and OO structure
- * \warning This program is a poor implementation - C style - and does not
- * utilize any of the C++ STL features.
  */
+#include <algorithm>  // test code
+#include <cassert>
 #include <iostream>
+#include <memory>
+#include <queue>
+#include <vector>
 
-struct node {
-    int val;
-    node *left;
-    node *right;
-};
+/**
+ * @namespace data_structure
+ * @brief Implementations of various data structures
+ */
+namespace data_structure {
+/**
+ * @namespace data_structure::BST
+ * @brief Implementation of Binary Search Tree data structure
+ */
+namespace BST {
 
-struct Queue {
-    node *t[100];
-    int front;
-    int rear;
+/** An struct node type.
+ *  Define a element for tree insert/remove operation
+ */
+struct node {
+    int val{};                   /**< int value of a node struct*/
+    std::shared_ptr<node> left;  /**< left subtree pointer */
+    std::shared_ptr<node> right; /**< right subtree pointer */
+    //~node() { std::cout << "node: destructor, val = " << val << '\n'; }
 };
 
-Queue queue;
-
-void enqueue(node *n) { queue.t[queue.rear++] = n; }
-
-node *dequeue() { return (queue.t[queue.front++]); }
+/** insert a node to tree
+ * with greater value to right subtree, smaller value node to left subtree
+ * \param[in] root root node of a tree
+ * \param[in] x a node with value to be insert
+ */
+void Insert(node* root, int x) {
+    if (root == NULL) {
```

Comment:
please consider the changes for the insert function which i suggested earlier.

Reply:
yes.i try early.but auto format checker failed.
you can get in commit:
adb42e4

so.i revert to last good code.

Reply:
@liushubin-gitHub what was the actual error? is removing the optimised code is only the solution?

Reply:
```
if (root == nullptr) {
    root = std::shared_ptr<node>(new node);
    root->val = x;
    root->left = nullptr;
    root->right = nullptr;
   return;
}